packages/language-server/src/hover.ts

```diff
@@ -0,0 +1,255 @@
+  astNode: XMLElement;
+}
+interface XMLKeyAttributeHover {
+  kind: "XMLKeyAttributeHover";
+  astNode: XMLAttribute;
+}
+interface XMLValueAttributeHover {
+  kind: "XMLValueAttributeHover";
+  astNode: XMLAttribute;
+}
+export type XMLHoverContext =
+  | XMLElementHover
+  | XMLKeyAttributeHover
+  | XMLValueAttributeHover;
+
+class HoverContextVisitor implements XMLAstVisitor {
+  public hoverContext: XMLHoverContext | undefined;
+  constructor(public chosenOffset: number | undefined) {}
+  visitXMLElement(node: XMLElement): void {
+    const openName = node.syntax.openName;
+    const closeName = node.syntax.closeName;
+    if (this.isInsideRange(openName)) {
+      this.hoverContext = { astNode: node, kind: "XMLElementHover" };
+    }
+
+    if (this.isInsideRange(closeName)) {
+      this.hoverContext = { astNode: node, kind: "XMLElementHover" };
+    }
+  }
+
+  visitXMLAttribute?(node: XMLAttribute): void {
+    const key = node.syntax.key;
+    const value = node.syntax.value;
+
+    if (this.isInsideRange(key)) {
+      this.hoverContext = { astNode: node, kind: "XMLKeyAttributeHover" };
+    }
+
+    if (this.isInsideRange(value)) {
+      this.hoverContext = { astNode: node, kind: "XMLValueAttributeHover" };
+    }
+  }
+
+  isInsideRange(name: XMLToken | undefined): boolean {
+    if (
+      name !== undefined &&
+      this.chosenOffset !== undefined &&
+      name.startOffset <= this.chosenOffset &&
+      name.endOffset >= this.chosenOffset
+    ) {
+      return true;
+    }
+
+    return false;
+  }
+}
+
+export function getHoverContext(
+  model: UI5SemanticModel,
+  textDocumentPosition: TextDocumentPositionParams,
+  document: TextDocument
+): Hover | undefined {
+  const documentText = document.getText();
+  const { cst, tokenVector } = parse(documentText);
+  const ast = buildAst(cst as DocumentCstNode, tokenVector);
+  const offset = document.offsetAt(textDocumentPosition.position);
+  const ui5Node = getUI5Node(ast, offset, model);
+  if (ui5Node != undefined) {
+    return transformToLspHover(ui5Node, model);
+  }
+
+  return undefined;
+}
+
+function getUI5Node(
+  ast: XMLDocument,
+  offset: number,
+  model: UI5SemanticModel
+): BaseUI5Node | undefined {
+  const visitor = new HoverContextVisitor(offset);
+  accept(ast, visitor);
+  if (visitor.hoverContext?.kind === undefined) {
+    return undefined;
+  }
+  switch (visitor.hoverContext?.kind) {
+    case "XMLElementHover":
+      return getUI5NodeByElement(visitor.hoverContext.astNode, model);
+    case "XMLKeyAttributeHover":
+      return getUI5NodeByKey(visitor.hoverContext.astNode, model);
+    case "XMLValueAttributeHover":
+      return getUI5NodeByValue(visitor.hoverContext.astNode, model);
+    case undefined:
+      return undefined;
+  }
+}
+
+function transformToLspHover(
+  ui5Node: BaseUI5Node,
+  model: UI5SemanticModel
+): Hover | undefined {
+  const hoverItem: Hover = {
+    contents: getNodeDocumentation(ui5Node, model),
```

Comment:
**Future:** add the range information (this should be returned from the visitor, I think, and transformed to LSP Range here)